### PR TITLE
Earcut: Upgrade to version 2.2.4.

### PR DESCRIPTION
**Pull request by yanranxiaoxi**
Upgrading the `Earcut` algorithm to the latest version (`2.2.4`) to fix infinite loop.

> Improved earcut performance by 10–15%.

### CODE REVIEW

**Comment by Mugen87** [comment]
If you want to upgrade the earcut implementation, it should not be necessary to update E2E screenshots. That only indicates something has changed or even broken.

**Comment by yanranxiaoxi** [comment]
> If you want to upgrade the earcut implementation, it should not be necessary to update E2E screenshots. That only indicates something has changed or even broken.

Thanks for your reply. I have also found this problem and am now testing it by myself.

File: src/extras/Earcut.js [hunk 1]
```diff
@@ -362,17 +374,20 @@ function compareX( a, b ) {
...
-		filterPoints( outerNode, outerNode.next );
-		filterPoints( b, b.next );
+		return outerNode;
 
 	}
 
+	const bridgeReverse = splitPolygon( bridge, hole );
+
+	// filter collinear points around the cuts
+	return filterPoints( bridgeReverse, bridgeReverse.next ); // there is different from the implementation of Earcut v2.2.4
```
Comment (Mugen87) [inline src/extras/Earcut.js:388]: Why did you change the earcut implementation here? If there is a bug, wouldn't it be better to report it to the Earcut repository instead?

Comment (Mugen87) [reply src/extras/Earcut.js:388]: To clarify my initial post: A broken E2E test potentially indicates a bug. That's why I have asked to verify why the output has changed. However, if the output is as intended (because something has been fixed/improved in the earcut code), it's not necessary to restore the previous behavior. In this case, it's fine to update the screenshot.

It's just important to know, why the output has changed.

Comment (yanranxiaoxi) [reply src/extras/Earcut.js:388]: I tagged here for assistance because there might be a problem (but it still returned the triangulation when I tested Earcut).

I'm feel so sorry about I did misunderstand what you meant yesterday.

I encountered an infinite loop problem in a particular model when I was learning the 3D engine, so I wanted to try updating the referenced earcut library to solve the problem. I'll reread the code until I know what's going on. Thank you for your review.

Comment (Mugen87) [reply src/extras/Earcut.js:388]: > I encountered an infinite loop problem in a particular model when I was learning the 3D engine

If you have the time it might be interesting to demonstrate the issue with a live example: https://jsfiddle.net/f2Lommf5/

**Comment by yanranxiaoxi** [comment]
Because Earcut v2.2.4 changed the logic of the simplified holes, so I updated webgl_modifier_tessellation.jpg here.

**Review by yanranxiaoxi — COMMENTED** [review]
Because Earcut v2.2.4 changed the logic of the simplified holes, so I updated webgl_modifier_tessellation.jpg here.

**Review by Mugen87 — APPROVED** [review]
